src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  // Figure out address of dlclose.
+  OUTCOME_TRY(address_dlclose, FindFunctionAddress(pid, "dlclose", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee.
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, kScratchPadSize));
+
+  // We want to do the following in the tracee:
+  // dlclose(handle);
+  // The calling convention is to put the parameter in registers rdi. Then we load the address of
+  // address_dlclose into rax and do the call. Assembly in Intel syntax (destination first), machine
+  // code on the right:
+
+  // movabs rax, handle               48 b8 handle
+  // mov rdi, rax                     48 8b f8
+  // movabs rax, address_dlclose      48 b8 address_dlclose
+  // call rax                         ff d0
+  // int3                             cc
+  MachineCode code;
+  code.AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(absl::bit_cast<uint64_t>(handle))
+      .AppendBytes({0x48, 0x8b, 0xf8})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_dlclose)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, kScratchPadSize);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  if (GetReturnValueOrDie(pid) != 0) {
+    FATAL("Unable to unload dynamic library from tracee.");
+  }
+
+  // Cleanup memory and registers.
+  RestoreRegistersOrDie(original_registers);
+  FreeMemoryOrDie(pid, address_code, kScratchPadSize);
+  return outcome::success();
+}
+
+ErrorMessageOr<uint64_t> FindFunctionAddress(pid_t pid, std::string_view function,
+                                             std::string_view module) {
+  auto modules = ReadModules(pid);
+  if (modules.has_error()) {
+    return modules.error();
+  }
+
+  std::string module_file_path;
+  uint64_t module_base_address = 0;
+  for (const auto& m : modules.value()) {
+    if (absl::StrContains(m.name(), module)) {
+      module_file_path = m.file_path();
+      module_base_address = m.address_start();
+    }
+  }
+  if (module_file_path.empty()) {
+    return ErrorMessage(absl::StrFormat("There is no module %s in process %u.", module, pid));
+  }
+
+  auto elf_file = ElfFile::Create(module_file_path);
+  if (elf_file.has_error()) {
+    return elf_file.error();
+  }
+
+  auto syms = elf_file.value()->LoadSymbolsFromDynsym();
+  if (syms.has_error()) {
+    return ErrorMessage(absl::StrFormat("Failed to load symbols for module %s: %s", module,
+                                        syms.error().message()));
+  }
+
+  auto it = syms.value().symbol_infos().begin();
+  while (it != syms.value().symbol_infos().end()) {
+    if (absl::StrContains(it->name(), function)) {
```

Comment:
I think in this case we should lookup by exact name. I see now why it works. This is because libc has a private symbol which contains dlopen but is not dlopen.